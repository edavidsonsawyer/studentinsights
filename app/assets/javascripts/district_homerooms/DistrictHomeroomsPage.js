import PropTypes from 'prop-types';
import React from 'react';
import _ from 'lodash';
import {shortSchoolName} from '../helpers/PerDistrict';
import GenericLoader from '../components/GenericLoader';
import ExperimentalBanner from '../components/ExperimentalBanner';
import SectionHeading from '../components/SectionHeading';
import School from '../components/School';
import {apiFetchJson} from '../helpers/apiFetchJson';
import {sortByGrade} from '../helpers/SortHelpers';
import {sortSchoolSlugsByGrade} from '../helpers/PerDistrict';


// Show homerooms at each grade level
export default function DistrictHomeroomsPage(props) {
  const url = '/api/district/homerooms_json';
  return (
    <div className="DistrictHomeroomsPage">
      <ExperimentalBanner />
      <GenericLoader
        promiseFn={() => apiFetchJson(url)}
        render={json => (
          <DistrictHomeroomsView
            districtName={json.district_name}
            students={json.students}
          />
        )}
      />
    </div>
  );
}

export class DistrictHomeroomsView extends React.Component {
  render() {
    const {districtKey} = this.context;
    const {students, districtName} = this.props;

    const grades = _.uniq(students.map(student => student.grade)).sort(sortByGrade);
    const sortedSchools = _.uniqBy(students.map(student => student.school), 'id').sort((a, b) => {
      const bySchool = sortSchoolSlugsByGrade(districtKey, a.slug, b.slug);
      if (bySchool !== 0) return bySchool;
      return computeStudentCountForSchool(a, students) - computeStudentCountForSchool(b, students);
    });
    // const totalCount = enrollments.reduce((count, enrollment) => count + enrollment.enrollment, 0);

    return (
      <div>
        <SectionHeading>Classrooms in {districtName}</SectionHeading>
        <table style={styles.table}>
          <thead>
            <tr>
              <th style={styles.cell}>&nbsp;</th>
              {sortedSchools.map(school => {
                return <th key={school.id} style={styles.cell}>
                  <School id={school.id} name={shortSchoolName(districtKey, school.local_id)} />
                </th>;
              })}
            </tr>
          </thead>
          <tbody>
            {grades.map(grade => {
              // const studentCountForGrade = students.filter(student => student.grade === grade);
              return (
                <tr key={grade}>
                  <td key="grade" style={styles.cell}>{grade}</td>
                  {sortedSchools.map(school => {
                    const studentsInGradeAtSchool = students
                      .filter(student => student.school.id === school.id)
                      .filter(student => student.grade === grade);
                    const homerooms = _.uniqBy(_.compact(studentsInGradeAtSchool.map(student => student.homeroom)), 'id');
                    return (
                      <td key={school.id} style={{...styles.cell, verticalAlign: 'top'}}>
                        <div>
                          {homerooms.map(homeroom => {
                            const studentsInHomeroom = studentsInGradeAtSchool.filter(student => student.homeroom && student.homeroom.id === homeroom.id);
                            return (
                              <a
                                key={homeroom.id}
                                style={{display: 'block'}}
                                title={JSON.stringify(studentsInHomeroom, null, 2)}
                                href={`/homerooms/${homeroom.id}`}>{studentsInHomeroom.length}: {homeroom.educator ? homeroom.educator.full_name : homeroom.name}</a>
                            );
                          })}
                        </div>
                      </td>
                    );
                  })}
                </tr>
              );
            })}
          </tbody>
        </table>
      </div>
    );
  }

  // Render a horizontal bar.
  renderBar(studentCount, options = {}) {
    const scaleFactor = options.scaleFactor || 0.5;
    const percentage = (studentCount === 0)
      ? 0
      : studentCount * scaleFactor;
    const width = (percentage > 100)
      ? '100%'
      : percentage + '%';
    const padding = percentage === 0 ? 0 : 3;
    const backgroundColor = options.backgroundColor || (percentage > 100 ? '#666' : '#ccc');
    const color = percentage > 100 ? 'white' : 'black';
    const text = percentage === 0 ? '\u00A0' : studentCount;

    return <div style={{...styles.bar, padding, width, color, backgroundColor}}>{text}</div>;
  }
}
DistrictHomeroomsView.contextTypes = {
  districtKey: PropTypes.string.isRequired
};
DistrictHomeroomsView.propTypes = {
  districtName: PropTypes.string.isRequired,
  students: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    first_name: PropTypes.string.isRequired,
    last_name: PropTypes.string.isRequired,
    house: PropTypes.string,
    counselor: PropTypes.string,
    grade: PropTypes.string.isRequired,
    has_photo: PropTypes.bool.isRequired,
    school: PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      local_id: PropTypes.string.isRequired,
      slug: PropTypes.string.isRequired
    }).isRequired,
    homeroom: PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      educator: PropTypes.shape({
        id: PropTypes.number.isRequired,
        email: PropTypes.string.isRequired,
        full_name: PropTypes.string.isRequired
      })
    })
  })).isRequired
};


const styles = {
  caption: {
    fontSize: 12,
    margin: 10
  },
  table: {
    borderCollapse: 'collapse',
    margin: 10,
    tableLayout: 'fixed',
    width: '90%',
    border: '1px solid #eee'
  },
  cell: {
    padding: 5,
    textAlign: 'left',
    border: '1px solid #eee',
    overflow: 'hidden'
  },
  bar: {
    borderLeft: '1px solid #aaa',
    height: '100%'
  }
};


function computeStudentCountForSchool(school, students) {
  return students.filter(student => student.school.id === school.id).length;
}
