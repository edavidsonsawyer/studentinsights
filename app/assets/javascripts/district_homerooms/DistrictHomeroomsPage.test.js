import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import fetchMock from 'fetch-mock/es5/client';
import {withDefaultNowContext} from '../testing/NowContainer';
import PerDistrictContainer from '../components/PerDistrictContainer';
import DistrictHomeroomsPage, {DistrictHomeroomsView} from './DistrictHomeroomsPage';
import districtHomeroomsJson from './districtHomeroomsJson.fixture';

function testProps(props = {}) {
  return {
    ...props
  };
}

function testEl(props, context = {}) {
  const districtKey = context.districtKey || 'somerville';
  return withDefaultNowContext(
    <PerDistrictContainer districtKey={districtKey}>
      <DistrictHomeroomsPage {...props} />
    </PerDistrictContainer>
  );
}

beforeEach(() => {
  fetchMock.restore();
  fetchMock.get('/api/district/homerooms_json', districtHomeroomsJson);
});

it('renders without crashing', () => {
  const props = testProps();
  const el = document.createElement('div');
  ReactDOM.render(testEl(props), el);
});

it('renders everything after fetch', done => {
  const props = testProps();
  const el = document.createElement('div');
  ReactDOM.render(testEl(props), el);

  setTimeout(() => {
    expect($(el).find('table tbody tr').length).toEqual(8);
    done();
  }, 0);
});

describe('DistrictHomeroomsView', () => {
  it('pure component matches snapshot', () => {
    const districtKey = 'somerville';
    const tree = renderer
      .create(withDefaultNowContext(
        <PerDistrictContainer districtKey={districtKey}>
          <DistrictHomeroomsView
            students={districtHomeroomsJson.students}
            districtName={districtHomeroomsJson.district_name} />
        </PerDistrictContainer>
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});


