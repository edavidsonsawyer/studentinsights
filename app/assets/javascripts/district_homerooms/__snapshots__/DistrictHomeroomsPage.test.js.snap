// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DistrictHomeroomsView pure component matches snapshot 1`] = `
<div>
  <div
    className="SectionHeading"
    style={
      Object {
        "borderBottom": "1px solid #333",
        "padding": 10,
      }
    }
  >
    <h4
      style={
        Object {
          "color": "black",
          "display": "inline",
        }
      }
    >
      Classrooms in 
      Localhost Public Schools
    </h4>
  </div>
  <table
    style={
      Object {
        "border": "1px solid #eee",
        "borderCollapse": "collapse",
        "margin": 10,
        "tableLayout": "fixed",
        "width": "90%",
      }
    }
  >
    <thead>
      <tr>
        <th
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          Â 
        </th>
        <th
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <a
            className="School"
            href="/schools/1"
            style={Object {}}
          >
            Brown
          </a>
        </th>
        <th
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <a
            className="School"
            href="/schools/6"
            style={Object {}}
          >
            West
          </a>
        </th>
        <th
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <a
            className="School"
            href="/schools/2"
            style={Object {}}
          >
            Healey
          </a>
        </th>
        <th
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <a
            className="School"
            href="/schools/9"
            style={Object {}}
          >
            Somerville High
          </a>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          KF
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[]"
          />
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[]"
          />
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[
  {
    \\"id\\": 1,
    \\"name\\": \\"HEA 003\\",
    \\"educator\\": {
      \\"id\\": 2,
      \\"email\\": \\"vivian@demo.studentinsights.org\\",
      \\"full_name\\": \\"Teacher, Vivian\\"
    }
  }
]"
          >
            <a
              href="/homerooms/1"
              style={
                Object {
                  "display": "block",
                }
              }
            >
              Teacher, Vivian
            </a>
          </div>
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[]"
          />
        </td>
      </tr>
      <tr>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          1
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[]"
          />
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[]"
          />
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[]"
          />
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[]"
          />
        </td>
      </tr>
      <tr>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          3
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[]"
          />
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[]"
          />
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[]"
          />
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[]"
          />
        </td>
      </tr>
      <tr>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          5
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[
  {
    \\"id\\": 3,
    \\"name\\": \\"WSNS 501\\",
    \\"educator\\": {
      \\"id\\": 7,
      \\"email\\": \\"marcus@demo.studentinsights.org\\",
      \\"full_name\\": \\"Teacher, Marcus\\"
    }
  }
]"
          >
            <a
              href="/homerooms/3"
              style={
                Object {
                  "display": "block",
                }
              }
            >
              Teacher, Marcus
            </a>
          </div>
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[
  {
    \\"id\\": 3,
    \\"name\\": \\"WSNS 501\\",
    \\"educator\\": {
      \\"id\\": 7,
      \\"email\\": \\"marcus@demo.studentinsights.org\\",
      \\"full_name\\": \\"Teacher, Marcus\\"
    }
  }
]"
          >
            <a
              href="/homerooms/3"
              style={
                Object {
                  "display": "block",
                }
              }
            >
              Teacher, Marcus
            </a>
          </div>
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[
  {
    \\"id\\": 2,
    \\"name\\": \\"HEA 500\\",
    \\"educator\\": {
      \\"id\\": 6,
      \\"email\\": \\"sarah@demo.studentinsights.org\\",
      \\"full_name\\": \\"Teacher, Sarah\\"
    }
  }
]"
          >
            <a
              href="/homerooms/2"
              style={
                Object {
                  "display": "block",
                }
              }
            >
              Teacher, Sarah
            </a>
          </div>
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[]"
          />
        </td>
      </tr>
      <tr>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          8
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[]"
          />
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[]"
          />
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[]"
          />
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[]"
          />
        </td>
      </tr>
      <tr>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          9
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[]"
          />
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[]"
          />
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[]"
          />
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[
  {
    \\"id\\": 5,
    \\"name\\": \\"SHS 942\\",
    \\"educator\\": {
      \\"id\\": 10,
      \\"email\\": \\"jodi@demo.studentinsights.org\\",
      \\"full_name\\": \\"Teacher, Jodi\\"
    }
  },
  {
    \\"id\\": 6,
    \\"name\\": \\"SHS 917\\",
    \\"educator\\": {
      \\"id\\": 12,
      \\"email\\": \\"bill@demo.studentinsights.org\\",
      \\"full_name\\": \\"Teacher, Bill\\"
    }
  }
]"
          >
            <a
              href="/homerooms/5"
              style={
                Object {
                  "display": "block",
                }
              }
            >
              Teacher, Jodi
            </a>
            <a
              href="/homerooms/6"
              style={
                Object {
                  "display": "block",
                }
              }
            >
              Teacher, Bill
            </a>
          </div>
        </td>
      </tr>
      <tr>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          10
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[]"
          />
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[]"
          />
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[]"
          />
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[
  {
    \\"id\\": 4,
    \\"name\\": \\"SHS ALL\\"
  }
]"
          >
            <a
              href="/homerooms/4"
              style={
                Object {
                  "display": "block",
                }
              }
            >
              SHS ALL
            </a>
          </div>
        </td>
      </tr>
      <tr>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          12
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[]"
          />
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[]"
          />
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[]"
          />
        </td>
        <td
          style={
            Object {
              "border": "1px solid #eee",
              "padding": 5,
              "textAlign": "left",
            }
          }
        >
          <div
            title="[]"
          />
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
